crates/bevy_ecs/src/entity/mod.rs

```diff
@@ -582,20 +631,82 @@ pub struct Entities {
     /// [`flush`]: Entities::flush
     pending: Vec<u32>,
     free_cursor: AtomicIdCursor,
-    /// Stores the number of free entities for [`len`](Entities::len)
-    len: u32,
+
+    /// This functions exactly like [`Self::meta`], only it counts backwards instead of forwards.
+    remote_entities: Vec<EntityMeta>,
+    /// The counter for [`Self::remote_entities`]. This will equal it's length after flushing.
+    remote_reservations: RemoteEntityReserver,
+}
+
+/// Provides remote reservation of an entity from anywhere.
+#[derive(Debug, Clone)]
+pub struct RemoteEntityReserver(Arc<RemoteIdCursor>);
+
+impl RemoteEntityReserver {
+    /// We do this because [`Entity::PLACEHOLDER`] has index of `u32::MAX` and generation of 1.
+    /// By starting at 2, we ensure we don't conflict with that, and, we allow similar Entity constants to be made in the future.
+    const REMOTE_FIRST_GENERATION: NonZero<u32> = const {
+        // SAFETY: We pass 2, which is greater than 0.
+        unsafe { NonZero::new_unchecked(2) }
```

Comment:
If we want to get rid of this unsafe, I think just doing `NonZero::new(2).unwrap()` should work? Both `new` and `unwrap` are const.